### PR TITLE
Minor bug fix for Array.Sort: Replace keys.Length with length for FloorLog2PlusOne

### PR DESCRIPTION
**Pull request by nietras**
During my work with implementing https://github.com/dotnet/corefx/issues/15329 I found an issue with the C# implementation of Array.Sort.

The wrong length is used for the introspective sort depth limit. It uses the whole arrays length instead of the sub view input.

The actual impact for this is hard to determine, but worst case the depth limit is way too high causing degenerate cases not to fall back to heap sort, when it should. As far as I can tell.

cc: @jkotas

### CODE REVIEW

**Comment by jkotas** [comment]
cc @joperezr @AlexGhiondea

File: src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs [hunk 1]
```diff
@@ -213,7 +213,7 @@ internal static void IntrospectiveSort(T[] keys, int left, int length, Compariso
             if (length < 2)
                 return;
 
-            IntroSort(keys, left, length + left - 1, 2 * IntrospectiveSortUtilities.FloorLog2PlusOne(keys.Length), comparer);
```
Comment (jkotas) [inline src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs]: The same bug seems to be in the non-generic sort version in Array.cs. Could you please fix it there as well?

Comment (nietras) [reply src/mscorlib/src/System/Collections/Generic/ArraySortHelper.cs]: @jkotas done, hope I've got them all now :)

**Comment by joperezr** [comment]
LGTM. thanks a lot for the help @nietras

**Comment by nietras** [comment]
@jkotas  is this fix not part of .NET Core 2.1 Preview1?

**Comment by nietras** [comment]
@jkotas I should note that the reason I ask is because this does have on impact on the result of the sorting of keys with items (values). Since `HeapSort` (for `IComparable` and for basic types where items type is not same as keys type) is then called in cases where it wasn't called before, hence new Span version calls `HeapSort` where .NET Core preview1 does not since it does not have the fix it seems. And `HeapSort` has an issue with how identical keys and their values are sorted.

**Comment by jkotas** [comment]
This fix is not in .NET Core 2.1 preview1
https://github.com/dotnet/coreclr/tree/release/2.1/src/mscorlib/src/System/Collections/Generic

**Pull request merged**